contracts/core/EntryPoint.sol

```diff
@@ -227,37 +228,50 @@ contract EntryPoint is IEntryPoint, StakeManager {
+     *          if the wallet has an aggregator that does not exist in the list, then
+     *          the call reverts with UnverifiedSignatureAggregator
      * @return preOpGas total gas used by validation (including contract creation)
      * @return prefund the amount the wallet had to prefund (zero in case a paymaster pays)
      * @return deadline until what time this userOp is valid (the minimum value of wallet and paymaster's deadline)
-     * @return actualAggregator the aggregator used by this userOp. if a non-zero aggregator is returned, the bundler must get its params using
-     *      aggregator.
+     * @return aggregator the aggregator returned by this wallet (or address(0), if it doesn't have one).
+     *          and only if it appears in the allowedAggregators list.
      * @return sigForUserOp - only if has actualAggregator: this value is returned from IAggregator.validateUserOpSignature, and should be placed in the userOp.signature when creating a bundle.
-     * @return sigForAggregation  - only if has actualAggregator:  this value is returned from IAggregator.validateUserOpSignature, and should be passed to aggregator.aggregateSignatures
-     * @return offChainSigInfo - if has actualAggregator, and offChainSigCheck is true, this value should be used by the off-chain signature code (e.g. it contains the sender's publickey)
+     * @return sigForAggregation - only if has actualAggregator:  this value is returned from IAggregator.validateUserOpSignature, and should be passed to aggregator.aggregateSignatures
      */
-    function simulateValidation(UserOperation calldata userOp, bool offChainSigCheck)
-    external returns (uint256 preOpGas, uint256 prefund, uint256 deadline, address actualAggregator, bytes memory sigForUserOp, bytes memory sigForAggregation, bytes memory offChainSigInfo) {
+    function simulateValidation(UserOperation calldata userOp, address[] memory allowedAggregators)
+    external returns (uint256 preOpGas, uint256 prefund, uint256 deadline, address aggregator, bytes memory sigForUserOp, bytes memory sigForAggregation) {
         uint256 preGas = gasleft();
 
         UserOpInfo memory outOpInfo;
 
-        (actualAggregator, deadline) = _validatePrepayment(0, userOp, outOpInfo, SIMULATE_NO_AGGREGATOR);
+        (aggregator, deadline) = _validatePrepayment(0, userOp, outOpInfo, SIMULATE_NO_AGGREGATOR);
         prefund = outOpInfo.prefund;
         preOpGas = preGas - gasleft() + userOp.preVerificationGas;
 
         numberMarker();
-        if (actualAggregator != address(0)) {
-            (sigForUserOp, sigForAggregation, offChainSigInfo) = IAggregator(actualAggregator).validateUserOpSignature(userOp, offChainSigCheck);
+        if (aggregator != address(0)) {
+            bool found = false;
+            if (allowedAggregators.length>0) {
+                for (uint i=0; i< allowedAggregators.length; i++) {
```

Comment:
I know I am coming late to this PR, but can I get a real quick TLDR why do we need to check the `aggregator` against the list of allowed ones in Solidity? Like, if this method is called in view mode and returns an aggregator address, what difference does it make if the check if the aggregator is allowed is done off-chain? 
Yes, you run some code of a non-whitelisted aggregator in a view mode then, but how much damage can this do?
Thanks!

Reply:
Previously the simulation did run the aggregator code, so we needed a way to prevent execution of unknown/unstaked aggregators to avoid DoS (e.g. aggregators that cause a lot of work, succeed in 1st simulation and fail the 2nd one in large quantities through a single storage slot change).  With the split to two steps, and the simulation only returning the aggregator and not executing it, this is no longer needed.  The bundler can look at the returned aggregator address and decides if it wants to proceed to the next step or drop the UserOp.  I think we can remove this.